### ISSUE TITLE
A possible bug in FeatureUtilities and issues on new SRL features in Edison

### ISSUE THREAD

**Issue by xinbowu2** [IllinoisCogComp/illinois-cogcomp-nlp#69]
@christos-c  @mssammon 
The snippet of codes containing the possible bug in FeatureUtilities : 

 `public static Set<Feature> conjoin(Set<Feature> feats1, Set<Feature> feats2) {

        if (feats1.size() == 0 || feats2.size() == 0)
            return new LinkedHashSet<>();

        Set<Feature> features = new LinkedHashSet<>(feats1.size() * feats2.size());

        for (Feature a : feats1) {
            for (Feature b : feats2) {

                if (a == b)    // the possible bug here. It compares the memory addresses of two features instead of the contents of the two features
                    features.add(a);
                else
                    features.add(a.conjoinWith(b));
            }
        }
        return features;
    }
`
See the bug as indicated in the snippet of codes. 

Issues on new SRL features in Edison
Because in one of our goup meetings, we have decided to use FeatureCollection for new SRL instead of CachedFeatureCollection which is used in old SRL, the problem above may raise errors in the new SRL. 
Because the old SRL uses CachedFeatureCollection, two identical features have the same memory addresses, the feature won't conjoin with itself. However, because the new SRL uses FeatureCollection without the caching functionality,  two identical features turn out to have different memory addresses will conjoin with itself. This leads to redundent features.

**Comment by mssammon** [comment]
Good catch. Will change this behavior in my current working branch. Thanks!

**Comment by xinbowu2** [comment]
Please make this change in my branch also, since it may determine which
implementation to be used for the new SRL.

2016-05-19 5:31 GMT+08:00 Mark Sammons <notifications@github.com>:

**Comment by mssammon** [comment]
After a second examination with Christos, it looks like the original behavior is correct after all.  It is checking whether two features are the same object; if they are two different objects with the same value, then conjoining them is legitimate. For example, a conjunction of two sequential pos tags might well get two labels "NN" and "NN" (for a pair of words such as "lamb chop"), in which case the conjunction operator should generate "NN&NN", as it currently does with the equality operator check.

**Comment by xinbowu2** [comment]
Ok, what behaviors do you expect for the new SRL on this?

**Comment by xinbowu2** [comment]
Is it ok to provide an options for each SRL feature extractor to cache features?

**Comment by mssammon** [comment]
Do you have an example where the behavior changes for SRL feature extraction when caching features, vs when features are not cached? If so, this might be a bug in feature caching.

**Comment by xinbowu2** [comment]
Take conjoin as an example: If we cache features for one constituent, when
we want  features for the same constituent again, the features will be
extracted from the cache, so the features just extracted have the same
memory address as the features extracted in the first time. An issue is
raised, when we conjoin two features. If the two features refer to the same
constituent and they have the same content, their memory address will be
the same and they won't conjoin with each other.
However, if we donot cache features, we may extract the same features for
the same constituent twice and they turn out to have different memory
addresses. So they will be conjoined together, even though they have the
same content and refer to the same constituent.

**Comment by mssammon** [comment]
Can you give me an example where you extract the same features for the same constituent twice, and they end up being conjoined together when we are not caching features?

**Comment by xinbowu2** [comment]
For example, we want to conjoin predicate features and a list of features
(it will be extracted from a feature collection extractor) and this
conjunction refers to only one single constituent. Both predicate features
and a list of features contain parse phrase type features. We will do as
the following codes indicated. The two feature extractors will extract
features sequentially. Predicate features will be firstly extracted, so the
parse phrase type features inside predicate features will be extracted
firstly. If we cache, the phrase type features will be put into the cache.
Then, the list of features will be extracted, but there is also a phrase
type features here, so if we cache, the feature will directly retrieved
from the cache instead of generating a new feature. When conjoining, if we
cache, two phrase type features from predicate features and the list of
features will have the same memory addresses, so they won't conjoin
together. However, if we donot cache, when we extract phrase type features
in the second time, it will be a completely new feature object, so when
conjoining, two phrase type features from predicate features and the list
of features will conjoin together, which turns out to behave differently
from the cached version.

2016-05-19 12:31 GMT+08:00 Mark Sammons <notifications@github.com>:

**Comment by xinbowu2** [comment]
FeatureUtilities.conjoin(new PredicateFeatures(),tmp) // tmp is a
feature extractor for a list of features.


2016-05-19 13:06 GMT+08:00 Jason Wu <17jasonwu@gmail.com>:

**Comment by xinbowu2** [comment]
I forget to mention another reason for old SRL won't conjoin the same
features is that both predicate features and tmp share the same cache. If
they donot share the same cache, the same features will still be conjoined.

2016-05-19 13:08 GMT+08:00 Jason Wu <17jasonwu@gmail.com>:

**Comment by mssammon** [comment]
Thank you for the detailed explanation.  I still think the code you flagged in Edison is fundamentally correct, because changing it leads to incorrect behavior in a very simple case. The fact that the cached behavior is different is problematic, but seems to require a different fix that can account for the case you describe. My intuition is that in actuality, this will not lead to a significant change in overall system behavior for SRL, so I don't think the fix is high priority, but we should find a solution that ensures cached feature behavior is the same as non-cached.

**Comment by xinbowu2** [comment]
You are right. So for now, is it ok to allow new SRL conjoin the same
features for the same constituent?

2016-05-19 21:09 GMT+08:00 Mark Sammons <notifications@github.com>:

**Comment by mssammon** [comment]
Yes, i think that's fine for now.
